internal/backends/python/python.go

```diff
@@ -347,7 +333,176 @@ func makePythonPoetryBackend(python string) api.LanguageBackend {
 	}
 }
 
-func listSpecfile() (map[api.PkgName]api.PkgSpec, error) {
+func ensureVenvExists(python string, venvPath string) (string, string) {
+	// Create venv if it does not exsit
+	if !util.Exists(venvPath) {
+		util.GetCmdOutputFallible([]string{
+			python, "-m", "venv", venvPath,
+		})
+	}
+	pipPath := path.Join(venvPath, "bin", "pip")
+	return venvPath, pipPath
+}
+
+// makePythonPipBackend returns a backend for invoking poetry, given an arg0 for invoking Python
+// (either a full path or just a name like "python3") to use when invoking Python.
+func makePythonPipBackend(python string) api.LanguageBackend {
+	var pipFlags []PipFlag
```

Comment:
Not sure what assumptions we're making here, but the way UPM is used is by executing a command and exiting, we won't have the luxury of using something stored in memory

Reply:
It was either this or add an additional `interface {}` parameter to every invocation of the `Add` function, since the pip backend needs information only derived during `ListSpecfile`. Colton and I kicked back and forth on a few options before we settled on this one, it works because `ListSpecfile` is always run before `Add`.

Perhaps I can also add a `len(...) > 0` check as well, to catch the case where it hadn't been run yet, or we just parse the specfile twice in a row and are OK with that, if you've got a preference I'm happy to hear it -- my preference would be to just add the `interface {}`, but that's a change with a high blast radius.